### ISSUE TITLE
imdb new format title page

### ISSUE THREAD

**Issue by Thomasdouscha** [tboothman/imdbphp#221]
Hello,

https://www.imdb.com/title/tt5013056/

imdbphp can not get page value of variables. Because it is new format page. This is serious issue! What should we do?

**Comment by duck7000** [comment]
Did you use the latest version of imdbphp? the one that includes the force old layout?

I did and this movie works fine for me
![Screenshot from 2021-07-01 21-57-17](https://user-images.githubusercontent.com/15360048/124182666-76af4380-dab7-11eb-8615-ee85df9ae5ef.png)

**Comment by Thomasdouscha** [comment]
3 times it couldnot and then it works. I have updated too many movies for testing. It works except 2 times. It seems that a little bit not stable.

**Comment by Thomasdouscha** [comment]
It was not like this before

**Comment by duck7000** [comment]
i agree that it is not stable, sometimes i have to refresh the search to get all results.

So it is sort of working but not for long i guess. 
If the new imdb layout stays then imdbphp will have to be adapted.. that will require a huge overhaul..

**Comment by GeorgeFive** [comment]
I'm using the force old layout option, and the only problem I've noticed is that country doesn't always return properly ( #220 ).  I assume it won't stay this way long though... ugh.

**Comment by GeorgeFive** [comment]
Fix to get episodes again:

/src/Imdb/Title.php
Line 791
Replace with:
            if (preg_match_all('|aria-label="(\d+) seasons"|Ui', $this->page["Title"],

**Comment by M0H4MM4D0UTIS** [comment]
Yes it is approved. I have this problem too
[This commit](https://github.com/tboothman/imdbphp/commit/e2cf0c8c8076d8719d1406d4bf5b6bb6e8f2e8d7) no longer works

**Comment by duck7000** [comment]
Maybe it's time to think about rewrite of imdbphp?

Thing is that it's not 100% sure that the new layout will stay, at this moment it only seems to effect the first page, if you click on trivia for example the old layout is shown (at least here in the Netherlands)
Same is true for cast and crew, technical specs etc.
It seems that all values that direct to a new page are shown in old layout, if that will stay that way is a big question..

**Comment by M0H4MM4D0UTIS** [comment]
I think it 's time to rewrite the system with xpath. In this case, large and small changes in the source can be easily adapted to the system

**Comment by duck7000** [comment]
@M0H4MM4D0UTIS 
True, i wrote a new method a while ago in xpath, much cleaner.

**Comment by M0H4MM4D0UTIS** [comment]
I am working on the code to adapt to the new version.
A pull request has been sent.
I will continue to work on part of the code. But I can not completely solve everything. Because my free time is not enough

**Comment by paxter** [comment]
We should thinking about forking this project. This project is not active maintained anymore. Quick fixes are essential for projects like this. No offense against the maintainers. But the project lacks activity and responses for months now.

Another possibility is to add some new maintainers.

**Comment by DaanDeJong2001** [comment]
@paxter one of the maintainers commented a few times here already. And project had 27 commit this year. Not that many, but certainly not a dead project. If you are skilled enough to think about forking, how about bringing in some PR to fix issues. Or are you just commenting on how other people should spent their time to help you?

@M0H4MM4D0UTIS could you comment why you are following the XPath track?
In all entries failing for me, I see all the relevant details in a structured schema.org block at the beginning of the file. Wouldn't it be much better to use this JSON block to read details? See https://github.com/tboothman/imdbphp/issues/218#issuecomment-850782349
As this is independent from layout, it is hopefully also more robust in case of layout changes.

**Comment by duck7000** [comment]
> We should thinking about forking this project. This project is not active maintained anymore. Quick fixes are essential for projects like this. No offense against the maintainers. But the project lacks activity and responses for months now.
> 
> Another possibility is to add some new maintainers.

well i second that, i have a few very hard discussions here in the past with the maintainers to get fixes and add on to the source
If somebody is willing to take over this project from izzysoft i would have expected that it would be maintained, but for a long time i have the feeling that the maintainers wait until someone comes up with a good idea/solution.

**Comment by M0H4MM4D0UTIS** [comment]
> @M0H4MM4D0UTIS could you comment why you are following the XPath track?
> In all entries failing for me, I see all the relevant details in a structured schema.org block at the beginning of the file. Wouldn't it be much better to use this JSON block to read details? See [#218 (comment)](https://github.com/tboothman/imdbphp/issues/218#issuecomment-850782349)
> As this is independent from layout, it is hopefully also more robust in case of layout changes.


The XPath method is much better, more efficient and cleaner than regex, and if the source format changes, corrections can be made in no time.
I used the JSON block in as many parts as possible. But the information is not complete. For example, it does not have a photo or a role for the actors

**Comment by duck7000** [comment]
Well i think that json will be a excellent option (how cool is that, directly from the source) but if the information is not there/not complete a fallback to Xpath is a good viable option, certainly much cleaner then regex.
regex is hard to follow and hard to maintain in my opinion

**Comment by duck7000** [comment]
> Fix to get episodes again:
> 
> /src/Imdb/Title.php
> Line 791
> Replace with:
> 
> ```
>             if (preg_match_all('|aria-label="(\d+) seasons"|Ui', $this->page["Title"],
>               $matches)) {
>                 $this->seasoncount = $matches[1][0];
>             }elseif (preg_match('/1 Season/', $this->page["Title"],
>               $matches)) {
>                 $this->seasoncount = 1;
>             } else {
>                 $this->seasoncount = 0;
>             }
> ```
> 
> Fix to get countries again:
> /src/Imdb/Title.php
> Line 1129
> Replace with:
> `if (preg_match_all('!/search\/title\/\?country_of_origin=[^>]+?>(.*?)<!m', $this->getPage("Title"),`

Can you report this as a bug or as a pull request? 
I'm also not getting any episodes or seasons anymore.

**Comment by Thomasdouscha** [comment]
> > Fix to get episodes again:
> > /src/Imdb/Title.php
> > Line 791
> > Replace with:
> > ```
> >             if (preg_match_all('|aria-label="(\d+) seasons"|Ui', $this->page["Title"],
> >               $matches)) {
> >                 $this->seasoncount = $matches[1][0];
> >             }elseif (preg_match('/1 Season/', $this->page["Title"],
> >               $matches)) {
> >                 $this->seasoncount = 1;
> >             } else {
> >                 $this->seasoncount = 0;
> >             }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Fix to get countries again:
> > /src/Imdb/Title.php
> > Line 1129
> > Replace with:
> > `if (preg_match_all('!/search\/title\/\?country_of_origin=[^>]+?>(.*?)<!m', $this->getPage("Title"),`
> 
> Can you report this as a bug or as a pull request?
> I'm also not getting any episodes or seasons anymore.

it works very well, thanks

**Comment by tboothman** [comment]
Probably worth agreeing on the style of the replacement code.
As has been discussed already the preference should be for structured metadata first - so the JSON LD, then using any computer readable attributes that indicate something is data (there used to be some itemProp attributes on elements e.g. `itemprop="episodeNumber"`) then id attributes or whatever's available.

My preference would be xpath over regular expressions as regex had a tendency to break on minor changes like some extra whitespace or an extra attribute to an element.

I don't think we should try to retain code that delt with the old layout - it'll just end up with confusing code that's unclear why it's there or if it does anything.

Fortunately while there are a very huge amount of methods in the Title class only ~30 use the title page so while this is a big effort it's not insurmountable. Nobody's going to (me included) fix all of this so I'd ask people contribute fixes (as PRs) to some methods they're interested in and hopefully we'll get to a point where the tests pass again. Thanks @M0H4MM4D0UTIS for making a start

**Comment by duck7000** [comment]
@tboothman 
Thanks for commenting on the issue.

@all
I'm allready trying to fix the recommendations method but i'm stuck and need help.
I don't understand how this jsonLD() works exactly, i can't get it working.
If i try with getPage it looks like i get json data in return as their are no html tags inside as far as i can see
If i try with a query like below i get a string that looks like json but i can't figure out what to do next.
it appears to be that the actual data is coming from below script tag, inside the actual html i can't seem to find any data.
I think this is the wrong approach?
```
$cells = $xp->query("//script[@id=\"__NEXT_DATA__\"]");
echo $cells{0}->nodeValue;
```
If somebody can help me get started it is much appreciated! and remember that i'm not a professional programmer, just a hobby.

**Comment by tboothman** [comment]
Wow there's a lot of data in that JSON. I wouldn't be sure that's meant to be there or will remain .. but it's probably a bit easier to parse than the html.

Probably the easiest way to show this is in the browser. If you go to https://www.imdb.com/title/tt0133093/ and open up the developer console and type this you'll see the recommendations
```javascript
JSON.parse(document.getElementById('__NEXT_DATA__').innerText).props.urqlState['1814075148'].data.title.moreLikeThisTitles.edges
```
I'm not entirely sure what 1814075148 is or how anything knows that number but it does vary between pages.

`<script id="__NEXT_DATA__"> {"some": "json"}</script>` is a html element with stringified JSON in it, so you need to find the element (`document.getElementById..` in the javascript), parse the json into objects (`JSON.parse()`) then traverse it to find the moreLikeThisTitles field.

In PHP that would be 
```php
$script = $xp->query("//script[@id=\"__NEXT_DATA__\"]")->item(0)->nodeValue;
$data = json_decode($script);
print_r($data->props->urqlState->{1814075148}->data->title->moreLikeThisTitles->edges);
```
The obvious problem here is you don't know what 1814075148 so you'd have to look through each one to find one with moreLikeThisTitles in it

**Comment by tboothman** [comment]
What might be interesting is looking into this graphapi that the frontend is using now.
I've just had a quick test and it doesn't use any auth so actually we should be able to get all the data we need from that rather than getting the html.

e.g. here's what it does when you ask for more info about the recommended film forrest gump on the matrix page:

```bash
curl 'https://api.graphql.imdb.com/' -H 'content-type: application/json' -H 'x-imdb-client-name: imdb-web-next' -H 'x-imdb-user-language: en-GB' -H 'x-imdb-user-country: GB' -H 'Origin: https://www.imdb.com' --data-raw '{"query":"query Title_Summary_Prompt($titleId: ID!) {\n  title(id: $titleId) {\n    plot {\n      plotText {\n        plainText\n        __typename\n      }\n      __typename\n    }\n    genres {\n      genres(limit: 3) {\n        text\n        __typename\n      }\n      __typename\n    }\n    principalCast: principalCredits(filter: {categories: [\"cast\"]}) {\n      ...Credits\n      __typename\n    }\n    principalCreators: principalCredits(filter: {categories: [\"creator\"]}) {\n      ...Credits\n      __typename\n    }\n    principalDirectors: principalCredits(filter: {categories: [\"director\"]}) {\n      ...Credits\n      __typename\n    }\n    __typename\n  }\n}\n\nfragment Credits on PrincipalCreditsForCategory {\n  category {\n    text\n    __typename\n  }\n  credits(limit: 3) {\n    name {\n      id\n      nameText {\n        text\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n","operationName":"Title_Summary_Prompt","variables":{"titleId":"tt0109830"}}'
```

```json
{
  "data": {
    "title": {
      "plot": {
        "plotText": {
          "plainText": "The presidencies of Kennedy and Johnson, the Vietnam War, the Watergate scandal and other historical events unfold from the perspective of an Alabama man with an IQ of 75, whose only desire is to be reunited with his childhood sweetheart.",
          "__typename": "Markdown"
        },
        "__typename": "Plot"
      },
      "genres": {
        "genres": [
          {
            "text": "Drama",
            "__typename": "Genre"
          },
          {
            "text": "Romance",
            "__typename": "Genre"
          }
        ],
        "__typename": "Genres"
      },
      "principalCast": [
        {
          "category": {
            "text": "Stars",
            "__typename": "CreditCategory"
          },
          "credits": [
            {
              "name": {
                "id": "nm0000158",
                "nameText": {
                  "text": "Tom Hanks",
                  "__typename": "NameText"
                },
                "__typename": "Name"
              },
              "__typename": "Cast"
            },
            {
              "name": {
                "id": "nm0000705",
                "nameText": {
                  "text": "Robin Wright",
                  "__typename": "NameText"
                },
                "__typename": "Name"
              },
              "__typename": "Cast"
            },
            {
              "name": {
                "id": "nm0000641",
                "nameText": {
                  "text": "Gary Sinise",
                  "__typename": "NameText"
                },
                "__typename": "Name"
              },
              "__typename": "Cast"
            }
          ],
          "__typename": "PrincipalCreditsForCategory"
        }
      ],
      "principalCreators": [],
      "principalDirectors": [
        {
          "category": {
            "text": "Director",
            "__typename": "CreditCategory"
          },
          "credits": [
            {
              "name": {
                "id": "nm0000709",
                "nameText": {
                  "text": "Robert Zemeckis",
                  "__typename": "NameText"
                },
                "__typename": "Name"
              },
              "__typename": "Crew"
            }
          ],
          "__typename": "PrincipalCreditsForCategory"
        }
      ],
      "__typename": "Title"
    }
  },
  "extensions": {
    "disclaimer": "Public, commercial, and/or non-private use of the IMDb data provided by this API is not allowed. For limited non-commercial use of IMDb data and the associated requirements see https://help.imdb.com/article/imdb/general-information/can-i-use-imdb-data-in-my-software/G5JTRESSHJBBHTGX#"
  }
}
```

**Comment by tboothman** [comment]
This is a lot easier than faffing around with html and weird json in a page:
![image](https://user-images.githubusercontent.com/1813336/124359529-5ac5b200-dc1d-11eb-8b2f-909f48a9c40c.png)

**Comment by M0H4MM4D0UTIS** [comment]
I implemented a method that allows you to use __NEXT_DATA__:
https://github.com/tboothman/imdbphp/pull/224/commits/e48ebe514ae954fa0c3837202890975751dee5fd


and sample using:
https://github.com/tboothman/imdbphp/pull/224/commits/a170175207d066cf0d41615a4376b785f2a0d52d

**Comment by duck7000** [comment]
@tboothman @M0H4MM4D0UTIS 
Thanks for the explaining and examples!
This looks promising, if this get merged i got a good starting point, thanks!

Maybe it will help others to rebuild the other methods with this info

The only thing missing is de info that is available if you click on the little i icon, this is a overlay and is not in the source or script tag.
Well no big deal after all.

**Comment by tboothman** [comment]
All the tests pass now thanks for @M0H4MM4D0UTIS #224 
These two methods almost definitely don't work but there are no tests for them:
mainPictures
prodNotes

**Comment by tboothman** [comment]
I've removed those two methods.
The usage of that NEXT_DATA json in get_episode_details and movie_recommendations has been removed as it sometimes returned incorrect results.
I think this about covers it so I'm going to make a release with this code.

**Comment by tboothman** [comment]
https://github.com/tboothman/imdbphp/releases/tag/v7.0.0
